source/validate_types.cpp

```diff
@@ -186,23 +186,133 @@ ModuleLayoutSection ValidationState_t::getLayoutStage() const {
 }
 
 void ValidationState_t::progressToNextLayoutStageOrder() {
-  module_layout_order_stage_ +=
-      module_layout_order_stage_ < GetModuleOrder().size();
-  if (module_layout_order_stage_ >= GetModuleOrder().size()) {
-    current_layout_stage_ = libspirv::ModuleLayoutSection::kFunction;
+  if (current_layout_stage_ <= GetModuleOrder().size()) {
+    current_layout_stage_ =
+        static_cast<ModuleLayoutSection>(current_layout_stage_ + 1);
   }
 }
 
 bool ValidationState_t::isOpcodeInCurrentLayoutStage(SpvOp op) {
-  const vector<SpvOp>& currentStage =
-      GetModuleOrder()[module_layout_order_stage_];
+  const vector<SpvOp>& currentStage = GetModuleOrder()[current_layout_stage_];
   return end(currentStage) != find(begin(currentStage), end(currentStage), op);
 }
 
-libspirv::DiagnosticStream ValidationState_t::diag(
-    spv_result_t error_code) const {
+DiagnosticStream ValidationState_t::diag(spv_result_t error_code) const {
   return libspirv::DiagnosticStream(
       {0, 0, static_cast<size_t>(instruction_counter_)}, diagnostic_,
       error_code);
 }
+
+Functions& ValidationState_t::get_functions() { return module_functions_; }
+
+bool ValidationState_t::in_function_body() const {
+  return module_functions_.in_function_body();
+}
+
+bool ValidationState_t::in_block() const {
+  return module_functions_.in_block();
+}
+
+Functions::Functions(ValidationState_t& module)
+    : module_(module), in_function_(false), in_block_(false) {}
+
+bool Functions::in_function_body() const { return in_function_; }
+
+bool Functions::in_block() const { return in_block_; }
+
+spv_result_t Functions::RegisterFunction(uint32_t id, uint32_t ret_type_id,
+                                         uint32_t function_control,
+                                         uint32_t function_type_id) {
+  assert(in_function_ == false &&
+         "Function instructions can not be declared in a function");
+  in_function_ = true;
+  id_.emplace_back(id);
+  type_id_.emplace_back(function_type_id);
+  declaration_type_.emplace_back(FunctionDecl::kFunctionDeclUnknown);
+  block_ids_.emplace_back();
+  variable_ids_.emplace_back();
+  parameter_ids_.emplace_back();
+
+  // TODO(umar): validate function type and type_id
+  (void)ret_type_id;
+  (void)function_control;
+
+  return SPV_SUCCESS;
+}
+
+spv_result_t Functions::RegisterFunctionParameter(uint32_t id,
+                                                  uint32_t type_id) {
+  assert(in_function_ == true &&
+         "Function parameter instructions can not be declared of a function");
+  (void)id;
+  (void)type_id;
```

Comment:
I am eventually going to check the parameter type and use the id in some way. This is done to avoid unused parameter warnings in clang.